_vendor/src/github.com/coreos/etcd/clientv3/lease.go

```diff
@@ -392,34 +392,45 @@ func (l *lessor) recvKeepAliveLoop() (gerr error) {
 		l.mu.Unlock()
 	}()
 
-	stream, serr := l.resetRecv()
-	for serr == nil {
-		resp, err := stream.Recv()
-		if err == nil {
-			l.recvKeepAlive(resp)
-			continue
-		}
-		err = toErr(l.stopCtx, err)
-		if err == rpctypes.ErrNoLeader {
-			l.closeRequireLeader()
-			select {
-			case <-time.After(retryConnWait):
-			case <-l.stopCtx.Done():
+	for {
+		stream, err := l.resetRecv()
+		if err != nil {
+			if canceledByCaller(l.stopCtx, err) {
 				return err
 			}
-		} else if isHaltErr(l.stopCtx, err) {
```

Comment:
The bug is here: when `recvKeepAliveLoop` meet a error it can't recognize, this goroutine exits.
If this goroutine is exited, `KeepAlive`  can't success, which is called by `NewSession`, lead to the latter fail.
As `NewSession` always fail, TiDB can't renewal its lease and stop to serve.